### PR TITLE
Added autologging safety validation exemptions, and fixed autolog does not log implicit batch_size for tf keras fit method

### LINKED ISSUE #5339: [BUG] tensorflow autolog doesn't capture batch size correct for tf Dataset input
### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **MLflow installed from (source or binary)**: pip?
- **MLflow version (run ``mlflow --version``)**: 1.23.1
- **Python version**: 3.8
- **npm version, if running the dev UI**: n/a
- **Exact command to reproduce**: see code below

### Describe the problem

When calling `fit` with a tensorflow dataset with a batch operation applied, the mlflow tracking ui records batch size of None, not the correct batch size. Note that [the fit docs](https://keras.io/api/models/model_training_apis/#fit-method) instruct to not supply `batch_size` when the input data is a dataset:
> batch_size: ... Do not specify the batch_size if your data is in the form of datasets, generators, or keras.utils.Sequence instances

The cause of this appears to be [this line](https://github.com/mlflow/mlflow/blob/1335164d20a6521dd4ebfe3f361d19fe9672efcf/mlflow/tensorflow/__init__.py#L860) in the patched fit function for auto-logging, which logs the batch_size param, regardless of whether the input to fit was a batched dataset.

### Code to reproduce issue

```python
import mlflow
import tensorflow as tf


def gen():
    for i in range(10):
        yield i, -i

dataset = tf.data.Dataset.from_generator(
    gen,
    output_signature=(
        tf.TensorSpec(shape=(), dtype=tf.float32),
        tf.TensorSpec(shape=(), dtype=tf.float32),
    )
)
dataset = dataset.batch(2)

mlflow.tensorflow.autolog()

model = tf.keras.Sequential()
model.add(tf.keras.layers.Dense(1, input_shape=(1,)))
model.compile(loss="mse")
model.fit(dataset, epochs=10)
```
### Other info / logs

N/A

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by MarkYHZhang**
Signed-off-by: Mark Zhang <mark.zhang@databricks.com>

## What changes are proposed in this pull request?

Fixes https://github.com/mlflow/mlflow/issues/5339 by providing implicit `batch_size` extraction for `tf.data.Dataset`, generator, and `tf.keras.utils.Sequence`. As part of this PR, I also added a way to add autologging safety validation exemptions.

## How is this patch tested?

Unit tests.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Autolog will now log implicit `batch_size` for Tensorflow Keras models with input data of types: `tf.data.Dataset`, generator, or `tf.keras.utils.Sequence`. 

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tensorflow/__init__.py
Comment (dbczumar) [inline mlflow/tensorflow/__init__.py:902]: For readability, can we name `args[0]`? e.g. `training_data = args[0]`?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:32]: Can we scope this to a particular index too?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:28]: Can we add a comment above this list explaining what it's for? I.e for specifying a list of function arguments that can be mutated by autologging code. We should also warn developers that this list shouldn't be modified unless absolutely necessary and that, if modified, a clear reason needs to be provided as an inline comment + thorough integration tests need to be introduced.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:857]: I don't think we need `nonlocal` here since we're not mutating these vars, right?

File: mlflow/tensorflow/__init__.py
Comment (apurva-koti) [inline mlflow/tensorflow/__init__.py:914]: What if someone passes in `training_data` via `kwargs`?

File: mlflow/tensorflow/__init__.py
Comment (apurva-koti) [inline mlflow/tensorflow/__init__.py:919]: Nit: don't need `first_batch`

File: mlflow/tensorflow/__init__.py
Comment (apurva-koti) [inline mlflow/tensorflow/__init__.py:921]: Shouldn't we refer to `training_data` instead of `args[0]`?

File: mlflow/utils/autologging_utils/safety.py
Comment (apurva-koti) [inline mlflow/utils/autologging_utils/safety.py:812]: can we use `mlflow.tensorflow.FLAVOR_NAME`

File: mlflow/utils/autologging_utils/__init__.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/__init__.py:596]: Nit: I think this should be named `is_iterator`, since the checks implemented are related to iterator attributes. This *is* the right thing to check, but we should adjust the naming.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:783]:
```suggestion
def _is_arg_exempt_from_validation(
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:801]:
```suggestion
    :param pos_arg_index: the index of the argument, if it is passed as a positional argument
    :param kwarg_name: the name of the argument, if it is passed as a keyword argument
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:955]: Rather than tracking the recursive validation stack depth and performing validation in the case where `depth==1`, I'd rather we pull the depth=0 validation procedure out of the `_validate` function, even if we have to repeat some code.

For example:

```
length_difference = len(autologging_call_args) - len(user_call_args)
assert (
    length_difference >= 0
), "{} expected inputs are missing from the call to the original function.".format(
    length_difference
)

for autologging_call_arg, user_call_arg in itertools.zip_longest(autologging_call_args, user_call_args):
    if not _is_arg_exempt_from_validation(...):
        _validate(autologging_call_arg, user_call_arg)

assert set(user_call_kwargs.keys()).issubset(set(autologging_call_kwargs.keys())), (
    "Keyword or dictionary arguments to original function omit"
    " one or more expected keys: '{}'".format(
        set(user_call_kwargs.keys()) - set(autologging_call_kwargs.keys())
    )
)

for key in autologging_call_kwargs.keys():
    if not _is_arg_exempt_from_validation(...):
        _validate(
            autologging_call_kwargs[key],
            user_call_kwargs.get(key, None),
        )
```

We can pull the logic for list / dict length and key subset validation into separate methods and reuse them here & in `_validate`.

Comment (dbczumar) [reply mlflow/utils/autologging_utils/safety.py:955]: Additionally, when validating positional args, can we pass `autologging_call_kwargs` to `_is_arg_exempt_from_validation`? This will tell us how to interpret the positional arg index, which can change based on whether or not an argument is specified as a kwarg.

For example, consider the following function accepting integer arguments:

```
def func(foo, bar):
```

if arg `foo` is exempt from validation and the method is called as `func(bar=1, foo=2)`, the current logic in `_is_arg_exempt_from_validation` will exempt `bar` *and* `foo`.

We should only exempt a positional arg if its name is not found in the set of kwargs.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:814]: I think this check will lead us to over-exempting positional args in the unlikely scenario where multiple adjacent args have the same type. See https://github.com/mlflow/mlflow/pull/5683/files#r851456203.

File: tests/autologging/test_autologging_safety_unit.py
Comment (dbczumar) [inline tests/autologging/test_autologging_safety_unit.py:1144]: Nit: Can we pass the strings `"autologging_integration_name"` and `"function_name"` as arguments, rather than empty strings, so that maintainers know how to interpret the purpose / meaning of these args?

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (dbczumar) [inline tests/tensorflow/test_tensorflow2_autolog.py:338]: Are there any other types of generators used with tf.keras that we should test?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:800]: Nit: In a function signature, an argument always has exactly one index. At call time, the function may be passed positionally at that index or may be passed via kwargs. This function refers to an argument's characteristics in the context of the function signature, so I think `arg_index` is a more appropriate name.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:817]: For testing purposes, can we move this to a more generic allowlist form like we had in the previous commits? I propose a list of named tuples with the following attributes:

1. autologging integration
2. function name
3. argument name
4. argument index
5. type_fn (a function that takes an object and returns True if the object matches the argument type, False otherwise)

File: tests/autologging/test_autologging_safety_unit.py
Comment (dbczumar) [inline tests/autologging/test_autologging_safety_unit.py:1405]: I think we're overfitting to the tensorflow case for unit testing argument exemption. In https://github.com/mlflow/mlflow/pull/5683/files#r851466125, I've proposed that we instead define exemptions as allowlist entries. Then, we can mock the allowlist here and verify that arguments are exempted properly for different cases (e.g. try a couple of different functions & types).

This way, we can also test that, e.g. if the type doesn't match or the function name doesn't match or the integration name doesn't match, arguments aren't exempted (negative testing).

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [reply mlflow/utils/autologging_utils/safety.py:800]: Nevermind, this refers to the index at which the argument was passed at call time. Can we explain that it's `None` if it's a keyword arg? I think `pos_arg_index` is a good name.

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (MarkYHZhang) [reply tests/tensorflow/test_tensorflow2_autolog.py:338]: There exist tf.keras built-in generators like: [ImageDataGenerator](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/ImageDataGenerator?version=nightly) and [TimeseriesGenerator](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/sequence/TimeseriesGenerator?version=nightly). But they all inherits from [Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence) or [Iterator](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/Iterator?version=nightly) (which also itself inherits from [Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence)) which are already being tested. Additionally, they will soon be deprecated in future versions:

> Prefer loading data with either [tf.keras.utils.text_dataset_from_directory](https://www.tensorflow.org/api_docs/python/tf/keras/utils/text_dataset_from_directory) or [tf.keras.utils.image_dataset_from_directory](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory), and then transforming the output [tf.data.Dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) with preprocessing layers.

[Source](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing?version=nightly)

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:794]: For the purpose of defining an allowlist, this should always be defined. It's the index in the function signature.

```suggestion
    argument_index: the index of the argument in the function signature.
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:796]:
```suggestion
    argument_name: the name of the argument in the function signature.
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:812]: Can we document this method?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:929]: This method name isn't quite accurate. It's asserting that the autologging call input length is `>=` the user call input length to make sure that we didn't delete user arguments.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:937]: This method name isn't quite accurate. The method is asserting that the autologging call input kwarg keys are a superset of the user call input kwarg keys to make sure that we didn't delete user keyword arguments.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:945]:
```suggestion
    def _assert_equivalent(autologging_call_input, user_call_input):
```

File: tests/autologging/test_autologging_safety_unit.py
Comment (dbczumar) [inline tests/autologging/test_autologging_safety_unit.py:1176]: @MarkYHZhang Can we apply https://github.com/mlflow/mlflow/pull/5683/files#diff-c18d334f6497b607d53017ccc716f80d1f2ff2cd989358c5bca64ae877f545ccR1146-R1148 to the rest of this file as well? I.e. instead of passing empty strings, include content that helps maintainers interpret the meaning of these args.

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:1010]: Positional arguments of a function call are always tuples. For simplicity, can we rewrite this logic with that assumption?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:1032]: The set of keyword arguments of a function call is always a dictionary. For simplicity, can we rewrite this logic with that assumption?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:1010]: The set of positional arguments of a function is always a tuple. For simplicity, can we rewrite this logic with that assumption?

File: tests/tensorflow/test_tensorflow2_autolog.py
Comment (dbczumar) [reply tests/tensorflow/test_tensorflow2_autolog.py:338]: Thanks!

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:803]:
```suggestion
    def matches(
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:807]:
```suggestion
        value,
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:800]:
```suggestion
    argument_index: int = None
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:801]:
```suggestion
    argument_name: str = None
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:811]: In the docstring, can we explain what this method does?

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:876]: Nice! This reads much better now :)

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:941]:
```suggestion
    def _assert_autologging_input_kwargs_are_superset(autologging_call_input, user_call_input):
```

File: mlflow/utils/autologging_utils/safety.py
Comment (dbczumar) [inline mlflow/utils/autologging_utils/safety.py:931]:
```suggestion
    def _assert_autologging_input_positional_args_are_superset(
```

**Review by dbczumar — APPROVED** [review]
LGTM once a few nitty naming comments are addressed. Thanks a bunch, @MarkYHZhang !

**Pull request merged**
